source/GameInterface/Services/MobileParties/Handlers/MobilePartyRecruitmentHandler.cs

```diff
@@ -0,0 +1,50 @@
+﻿using Common.Logging;
+using Common.Messaging;
+using GameInterface.Services.MobileParties.Messages;
+using GameInterface.Services.MobileParties.Patches;
+using GameInterface.Services.ObjectManager;
+using Serilog;
+using System.Reflection;
+using TaleWorlds.CampaignSystem;
+using TaleWorlds.CampaignSystem.CampaignBehaviors;
+using TaleWorlds.CampaignSystem.Party;
+using TaleWorlds.CampaignSystem.Settlements;
+
+namespace GameInterface.Services.MobileParties.Handlers
+{
+    /// <summary>
+    /// Handles all mobile party recruitment in game
+    /// </summary>
+    public class MobilePartyRecruitmentHandler : IHandler
+    {
+        private readonly IMessageBroker messageBroker;
+        private readonly IObjectManager objectManager;
+        private readonly ILogger Logger = LogManager.GetLogger<MobilePartyRecruitmentHandler>();
+        private RecruitmentCampaignBehavior recruitmentCampaignBehavior;
+
+        public MobilePartyRecruitmentHandler(IMessageBroker messageBroker, IObjectManager objectManager)
+        {
+            this.messageBroker = messageBroker;
+            this.objectManager = objectManager;
+            messageBroker.Subscribe<UnitRecruitGranted>(Handle);
+        }
+
+        public void Dispose()
+        {
+            messageBroker.Unsubscribe<UnitRecruitGranted>(Handle);
+        }
+
+        private void Handle(MessagePayload<UnitRecruitGranted> obj)
+        {
+            var payload = obj.What;
+
+            if (objectManager.TryGetObject(payload.PartyId, out MobileParty mobileParty) == false)
+            {
+                Logger.Error("Could not handle UnitRecruitGranted, PartyId not found: {id}", payload.PartyId);
+                return;
+            }
+
+            UnitRecruitPatch.RunOriginalAddToCounts(CharacterObject.Find(payload.CharacterId), payload.Amount, mobileParty, payload.IsPrisonRoster);
```

Comment:
Use object manager to get character object, find is slow

Reply:
There are not character objects in the object manager